data_structures/arrays/single_number.rb

```diff
@@ -0,0 +1,61 @@
+# Challenge name: Single Number
+#
+# Given a non-empty array of integers nums, every element appears twice
+# except for one. Find that single one.
+#
+# Follow up: Could you implement a solution with a linear runtime
+# complexity and without using extra memory?
+#
+# @param {Integer[]} nums
+# @return {Integer}
+
+# 
+# Approach 1: Hash map
+#
+# Time Complexity: O(n)
+#
+def single_number(nums)
+  result_hash = {}
+  nums.each do |num|
+    if result_hash[num]
+      result_hash[num] +=1 
+    else
+      result_hash[num] = 1
+    end
+  end
+
+  result_hash.each do |k, v|
+    return k if v == 1
+  end
+end
+
+nums = [2, 2, 1]
+puts(single_number(nums))
+# Output: 1
+nums = [4, 1, 2, 1, 2]
+puts(single_number(nums))
+# Output: 4
+nums = [1]
+puts(single_number(nums))
+# Output: 1
+
+#
+# Approach 2: Use Ruby .count()
+#
+# Time Complexity: O(1)
```

Comment:
Not sure this approach is O(1).

I think `Array#find` and `Array#count` iterate over an array and the time complexity of this approach will be O(n^2).

Reply:
Ohhh I see! Thank you, I am still new to time complexity and the notation. I will update this.

Reply:
Yeah @jsca-kwok behind the scenes the count and find do a loop.

Reply:
Count: https://github.com/ruby/ruby/blob/e019dd24df4ed7063ad80d4c2e4070141793f598/array.c#L6084